### PR TITLE
chore(telemetry): track configurations that set trace sampling rates

### PR DESCRIPTION
**Pull request by mabdinur**
Submits `DD_TRACE_SAMPLE_RATE` and `DD_TRACE_RATE_LIMIT` configurations to the instrumentation telemetry platform.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: tests/tracer/test_sampler.py [hunk 1]
```diff
@@ -664,11 +666,12 @@ def test_datadog_sampler_init():
...
         ], "DatadogSampler initialized with DD_TRACE_SAMPLE_RATE=0 envvar should hold sample_rate=0"
 
-    with override_env(dict(DD_TRACE_SAMPLE_RATE="invalid-rate")):
+
+    with override_global_config(dict(_trace_sample_rate="asdf")):
         with pytest.raises(ValueError):
             DatadogSampler()
 
-    with override_env(dict(DD_TRACE_RATE_LIMIT="invalid-limit")):
+    with override_global_config(dict(_trace_rate_limit="asdf")):
```
Comment (mabdinur) [inline tests/tracer/test_sampler.py:674]:
```suggestion
    with override_global_config(dict(_trace_rate_limit="invalid-limit")):
```

File: tests/tracer/test_sampler.py [hunk 2]
```diff
@@ -664,11 +666,12 @@ def test_datadog_sampler_init():
             SamplingRule(sample_rate=0)
         ], "DatadogSampler initialized with DD_TRACE_SAMPLE_RATE=0 envvar should hold sample_rate=0"
 
-    with override_env(dict(DD_TRACE_SAMPLE_RATE="invalid-rate")):
+
```
Comment (mabdinur) [inline tests/tracer/test_sampler.py:669]:
```suggestion
```

File: tests/tracer/test_sampler.py [hunk 3]
```diff
@@ -655,7 +656,8 @@ def test_datadog_sampler_init():
             SamplingRule(sample_rate=0.5)
         ], "DatadogSampler initialized with no arguments and envvars set should hold a sample_rate from the envvar"
 
-    with override_env(dict(DD_TRACE_SAMPLE_RATE="0")):
+    # DD_TRACE_SAMPLE_RATE=0
```
Comment (mabdinur) [inline tests/tracer/test_sampler.py:659]:
```suggestion
```

**Comment by mabdinur** [comment]
> ## Benchmarks
> Benchmark execution time: 2023-08-03 12:59:27
> 
> Comparing candidate commit [342ebbb](https://github.com/DataDog/dd-trace-py/commit/342ebbbf30c5eda8153aa4b47868ba298de73b9d) in PR branch `munir/telemetry-track-trace-sample-rates` with baseline commit [98a7d39](https://github.com/DataDog/dd-trace-py/commit/98a7d39eb1964fd1310c0ddcfdccf11eebe2bf5c) in branch `1.x`.
> 
> Found 4 performance improvements and 9 performance regressions! Performance is the same for 81 metrics, 0 unstable metrics.
> 
> #### scenario:flasksimple-baseline
> * 🟥 `max_rss_usage` [+701.243KB; +863.019KB] or [+2.101%; +2.586%]
> 
> #### scenario:flasksimple-iast-get
> * 🟥 `max_rss_usage` [+681.089KB; +854.092KB] or [+2.039%; +2.557%]
> 
> #### scenario:flasksimple-profiler
> * 🟥 `max_rss_usage` [+684.715KB; +818.517KB] or [+2.050%; +2.450%]
> 
> #### scenario:flasksimple-tracer
> * 🟥 `max_rss_usage` [+672.576KB; +833.933KB] or [+2.015%; +2.498%]
> 
> #### scenario:sethttpmeta-obfuscation-disabled
> * 🟩 `max_rss_usage` [-738.290KB; -585.947KB] or [-2.577%; -2.045%]
> 
> #### scenario:sethttpmeta-obfuscation-regular-case-explicit-query
> * 🟩 `max_rss_usage` [-861.180KB; -708.407KB] or [-2.975%; -2.447%]
> 
> #### scenario:sethttpmeta-obfuscation-regular-case-implicit-query
> * 🟥 `max_rss_usage` [+659.263KB; +807.924KB] or [+2.343%; +2.871%]
> 
> #### scenario:sethttpmeta-obfuscation-worst-case-explicit-query
> * 🟥 `max_rss_usage` [+663.555KB; +815.920KB] or [+2.355%; +2.896%]
> 
> #### scenario:span-add-tags
> * 🟥 `max_rss_usage` [+0.968MB; +1.106MB] or [+2.366%; +2.703%]
> 
> #### scenario:span-start-finish-telemetry
> * 🟩 `max_rss_usage` [-815.298KB; -646.155KB] or [-2.863%; -2.269%]
> 
> #### scenario:span-start-finish-traceid128
> * 🟩 `max_rss_usage` [-780.333KB; -636.064KB] or [-2.743%; -2.236%]
> 
> #### scenario:tracer-large
> * 🟥 `max_rss_usage` [+1.144MB; +1.312MB] or [+4.072%; +4.668%]
> 
> #### scenario:tracer-small
> * 🟥 `max_rss_usage` [+0.964MB; +1.126MB] or [+3.515%; +4.106%]

Telemetry payloads are not flushed when benchmarks are run. This increase in overhead is expected.

File: tests/tracer/test_sampler.py [hunk 4]
```diff
@@ -646,7 +647,7 @@ def test_datadog_sampler_init():
         SamplingRule(sample_rate=0.5)
     ], "DatadogSampler initialized with default_sample_rate should hold a SamplingRule with that rate"
 
-    with override_env(dict(DD_TRACE_SAMPLE_RATE="0.5", DD_TRACE_RATE_LIMIT="10")):
+    with override_global_config(dict(_trace_sample_rate="0.5", _trace_rate_limit=10)):
```
Comment (ZStriker19) [inline tests/tracer/test_sampler.py:650]: why are these test changes necessary?

Comment (mabdinur) [reply tests/tracer/test_sampler.py:650]: Previously the ``DD_TRACE_SAMPLE_RATE`` envar was always read when setting the sample rate. Now this environment variable is read once and it's value is stored in `config. _trace_sample_rate`. 

We need to override the global config object to run these tests.

Comment (ZStriker19) [reply tests/tracer/test_sampler.py:650]: Ah I see, got it! We'd basically need to run these in a subprocess, initializing a new tracer each time in order for the envar to work.

**Review by majorgreys — APPROVED** [review]

**Pull request merged**